### DISCUSSION TITLE
Mocking an external function in integration tests

### DISCUSSION THREAD

**Discussion by jgerrish** [asomers/mockall#350] [Questions]
I'm attempting to mock a function in an external crate in my library.  Mocking works fine in unit tests, but it's failing in integration tests.

I have a test case showing this behavior here:
https://github.com/jgerrish/mockall-mock-external-fn-in-integration-test

The external function I'm mocking is the crossterm read function, found in the namespace:

crossterm::event::read

We created a wrapper using automock and double in src/lib.rs for read in the crossterm and event modules, similar to the ffi example in the mockall library.  There is no associated struct.

This works for the unit tests.  But trying to move tests to integration tests fails.

The errors we're getting are either
`no mock_event` in `crossterm`
or
cannot find function `read_context` in module `event`

I found some cases to remove those errors, but the mocking fails and calls the normal crossterm library function.

I'm not sure if I'm failing to reference the correct namespace or module, or what the issue is.

Thank you for any help,

Josh

**Comment by asomers** [comment]
Easy problem. You did `#[cfg_attr(test, automock)]` in your src.  The `test` attribute is only true when building unit tests.  If you want to use the mocks from integration tests, then you should just do `#[automock]`.  Of course, that means the mocks will be included in your shipped library.  So, you may want to create a custom cfg variable, or a cargo feature, or something to control them.  Or do what I do and just move all mock-based testing into unit tests.

**Reply by jgerrish** [reply]
Thank you.  You're right, just using #[automock] generates the _context and mock_ functions.  Your description helped me understand the reason for that, thank you.

But the library doesn't use the mocked closure, instead calling the external function, but only in integration tests.  The repo has been updated to show this case.

I've moved my tests that rely on mocking to unit tests for now, per your suggestion.  That works.
